Source/Core/DolphinQt2/Config/GeckoWidget.cpp

```diff
@@ -0,0 +1,169 @@
+// Refer to the license.txt file included.
+
+#include "DolphinQt2/Config/GeckoWidget.h"
+
+#include <QGridLayout>
+#include <QListWidget>
+#include <QMessageBox>
+#include <QPushButton>
+#include <QTextEdit>
+
+#include "Common/FileUtil.h"
+#include "Core/ConfigManager.h"
+#include "Core/GeckoCodeConfig.h"
+#include "DolphinQt2/Config/CheatWarningWidget.h"
+#include "DolphinQt2/GameList/GameFile.h"
+
+constexpr int ENTRY_INDEX = Qt::UserRole;
+
+GeckoWidget::GeckoWidget(const GameFile& game)
+    : m_game(game), m_game_id(game.GetGameID().toStdString()), m_game_revision(game.GetRevision())
+{
+  CreateWidgets();
+  ConnectWidgets();
+
+  m_game_ini_default = SConfig::GetInstance().LoadDefaultGameIni(m_game_id, m_game_revision);
+  m_game_ini_local = SConfig::GetInstance().LoadLocalGameIni(m_game_id, m_game_revision);
+
+  m_gecko_codes = Gecko::LoadCodes(m_game_ini_default, m_game_ini_local);
+
+  UpdateList();
+}
+
+void GeckoWidget::CreateWidgets()
+{
+  auto* layout = new QGridLayout;
+  m_warning = new CheatWarningWidget(m_game_id);
+  m_code_list = new QListWidget;
+  m_code_description = new QTextEdit;
+  m_code_view = new QListWidget;
+  m_download_codes = new QPushButton(tr("Download Codes (WiiRD Database)"));
+
+  m_code_description->setTextInteractionFlags(Qt::TextBrowserInteraction);
+
+  m_download_codes->setEnabled(!m_game_id.empty());
+
+  layout->addWidget(m_warning, 0, 0);
+  layout->addWidget(m_code_list, 1, 0);
+  layout->addWidget(m_code_description, 2, 0);
+  layout->addWidget(m_code_view, 3, 0);
+  layout->addWidget(m_download_codes, 4, 0);
+
+  setLayout(layout);
+}
+
+void GeckoWidget::ConnectWidgets()
+{
+  connect(m_code_list, &QListWidget::itemSelectionChanged, this, &GeckoWidget::OnSelectionChanged);
+  connect(m_code_list, &QListWidget::itemChanged, this, &GeckoWidget::OnItemChanged);
+
+  connect(m_download_codes, &QPushButton::pressed, this, &GeckoWidget::DownloadCodes);
+
+  connect(m_warning, &CheatWarningWidget::OpenGeneralSettings, this,
+          &GeckoWidget::OpenGeneralSettings);
+}
+
+void GeckoWidget::OnSelectionChanged()
+{
+  auto items = m_code_list->selectedItems();
+
+  if (items.empty())
+    return;
+
+  auto* selected = items[0];
+
+  auto& code = m_gecko_codes[selected->data(ENTRY_INDEX).toInt()];
+
+  QString notes;
+
+  for (const std::string& line : code.notes)
+    notes.append(QString::fromStdString(line));
+
+  m_code_description->setText(
+      tr("Creator: %1\nDescription:\n\n%2")
+          .arg(QString::fromStdString(code.creator), notes.isEmpty() ? tr("N/A") : notes));
+
+  m_code_view->clear();
+
+  for (const auto& c : code.codes)
+    m_code_view->addItem(QString::asprintf("%08X  %08X", c.address, c.data));
+}
+
+void GeckoWidget::OnItemChanged(QListWidgetItem* item)
+{
+  // TODO: Implement Cheat Warning Panel
+
+  m_gecko_codes[item->data(ENTRY_INDEX).toInt()].enabled = (item->checkState() == Qt::Checked);
+  Gecko::SaveCodes(m_game_ini_local, m_gecko_codes);
```

Comment:
The INI-file should be re-read here, to pick up any changes that might have been written to it since the widget was created.